src/main/java/duckling/requests/BaseRequest.java

```diff
@@ -32,10 +34,27 @@ public BaseRequest(String baseRequest) {
 
         List<String> tokens = Arrays.asList(baseRequest.split(" "));
 
+        Function<String,String> parsePath = (path) -> {
+            try {
+                return path.split("\\?")[0];
+            } catch (ArrayIndexOutOfBoundsException exception) {
+                return "";
+            }
+        };
+
+        Function<String,String> parseQuery = (path) -> {
+            try {
+                return path.split("\\?")[1];
+            } catch (ArrayIndexOutOfBoundsException exception) {
+                return "";
+            }
+        };
+
```

Comment:
This constructor is getting pretty long at this point. Local functions are kind of cool, but there are a bunch of reasons it could be useful to split them out into methods. Besides maintainability, one potentially counterintuitive one is performance: with smaller methods, the JVM has more choices it can make about what to inline (see http://normanmaurer.me/blog/2014/05/15/Inline-all-the-Things/).

It looks like ultimately `parsePath` and `parseQuery` are doing the same work of path-splitting, with the same input (`tokens.get(PATH_INDEX)`). This makes me wonder if it might be worthwhile for some shared function/method to do the job of splitting, which could be done 1x, and then you could grab the path/query pieces as a layer on top?